### PR TITLE
Development of functional test scripts and issues found in development

### PR DESCRIPTION
**Pull request by jason-rodri**
This PR is based on the development of FIPS functional test scripts.

Script was developed to demonstrate that outstanding 186-5 feature were disabled  [adding 186-5 regression script](https://github.com/ciq-rocky-fips/openssl/pull/1/commits/dc139259ab5f92271eadb11b00f7616dab2d0710)

Remove hard coded module name, module name is not set in the configure command. This commit will also be merged into the 3.0.7-FIPS branch. [fixing issue with FIPS_VERSION and FIPS_NAME](https://github.com/ciq-rocky-fips/openssl/pull/1/commits/19b47d4e17f5f8dd40c783dc28d7c1377f140225)

Added functional code for inducing failure in the DH PCT. Code was not in the upstream so could not cherrypick. This update is needed to pass lab functional testing: [adding FT code to DH PCT](https://github.com/ciq-rocky-fips/openssl/pull/1/commits/0d04910b52824e426b9fd1b1e3f957cad699f3e1)

The openssl is used to exercise openssl functionality, so added the functional test controller to the app. [adding the ability to control functional tests to openssl app](https://github.com/ciq-rocky-fips/openssl/pull/1/commits/1ab95ddf4a7a76f62e97f0947954b679d4656aa1)

Since we are using the openssl app to drive functional testing we don't need the custom code so I will clean it up and add a script to exercise  the test cases.  [redsign of functional testing scripts](https://github.com/ciq-rocky-fips/openssl/pull/1/commits/a0fbbcd4c0df63cbc099ae20747de3d81edc82a6)
 
Updating build script [setting FIPS_VERSION and FIPS_name in build script](https://github.com/ciq-rocky-fips/openssl/pull/1/commits/44e8d85ea62bc740b4d68847ca5c854a47da0c0a)

### CODE REVIEW

File: providers/fips/fipsprov.c [hunk 1]
```diff
@@ -187,7 +187,7 @@ static int fips_get_params(void *provctx, OSSL_PARAM params[])
                                               &fips_prov_ossl_ctx_method);
 
     p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_NAME);
-    if (p != NULL && !OSSL_PARAM_set_utf8_ptr(p, "Rocky Enterprise Linux 9 - OpenSSL FIPS Provider"))
+    if (p != NULL && !OSSL_PARAM_set_utf8_ptr(p, ROCKY_FIPS_NAME))
```
Comment (jason-rodri) [inline providers/fips/fipsprov.c:190]: Replacing hard coded FIPS module name with a variable. The variable is set from the configure command

File: FT/186-5.regression.sh
Comment (jallisonciq) [inline FT/186-5.regression.sh:1]: Lines: 11, 18, 33, 37, 38 in file FT/186-5.regression.sh have trailing whitespace.

Can you fix ? (nit-picky I know, but it just looks cleaner).

**Review by jallisonciq — APPROVED** [review]

File: FT/186-5.regression.sh
Comment (jallisonciq) [reply FT/186-5.regression.sh:1]: I would just merge this into the patch that adds this (use git rebase --interactive with the "squash" command) to merge this into the previous patch. Doesn't need to be a separate commit.

**Review by jallisonciq — APPROVED** [review]

**Review by jallisonciq — COMMENTED** [review]
I don't understand what:
```
typedef struct st_args {
   int enable;
   int called;
} SELF_TEST_ARGS;

static SELF_TEST_ARGS self_test_args = { 0 };
```
is doing ? I don't see it referenced anywhere in this change. Am I missing something ?

**Review by jallisonciq — COMMENTED** [review]
The change in line 27 seems identical to the previous version except for the removal of whitespace. Can you put that space back as you're not really changing this line.

Cutting out whitespace changes reduces the number of diff lines needed to be reviewed.

File: apps/openssl.c [hunk 2]
```diff
@@ -231,6 +233,79 @@ static void setup_trace(const char *str)
...
+
+static OSSL_CALLBACK self_test_events;
+static char *self_test_corrupt_desc = NULL;
+static char *self_test_corrupt_type = NULL;
+static int self_test_log = 1;
+static int quiet = 0;
+
+static OSSL_PROVIDER *prov_null = NULL;
+static OSSL_LIB_CTX *libctx = NULL;
+static SELF_TEST_ARGS self_test_args = { 0 };
```
Comment (jason-rodri) [inline apps/openssl.c]: copy pasta

File: make_install.sh [hunk 3]
```diff
@@ -25,17 +24,19 @@ export HASHBANGPERL=/usr/bin/perl
 	--prefix=$cwd --openssldir=/etc/pki/tls $sslflags \
 	--system-ciphers-file=/etc/crypto-policies/back-ends/openssl.config \
 	zlib enable-camellia enable-seed enable-rfc3779 enable-sctp \
-	enable-cms enable-md2 enable-rc5 enable-ktls enable-fips \
-	no-mdc2 no-ec2m no-sm2 no-sm4 no-des no-dsa enable-buildtest-c++\
-	shared  $sslarch $RPM_OPT_FLAGS '-DDEVRANDOM="\"/dev/urandom\"" -DROCKY_FIPS_VERSION="\"$fips\""'\
-	-Wl,-rpath=$cwd/lib64 -Wl,--allow-multiple-definition --debug --openssldir=$cwd no-hw shared
+	enable-cms enable-md2 enable-rc5 enable-ktls enable-fips\
```
Comment (jallisonciq) [inline make_install.sh]: This looks like only a whitespace change.

**Review by ciq-sahlberg — APPROVED** [review]

File: apps/openssl.c [hunk 4]
```diff
@@ -231,6 +233,63 @@ static void setup_trace(const char *str)
...
+        fprintf(stderr, "%s ", phase);
+        goto err;
+    }
+end:
+    ret = 1;
+err:
+    return ret;
+}
+
+
```
Comment (jallisonciq) [inline apps/openssl.c]: Remove the two extra whitespace lines (290 and 291) please.

File: apps/openssl.c [hunk 5]
```diff
@@ -231,6 +233,63 @@ static void setup_trace(const char *str)
...
+static OSSL_CALLBACK self_test_events;
+static char *self_test_corrupt_desc = NULL;
+static char *self_test_corrupt_type = NULL;
+static int self_test_log = 1;
+
+static int self_test_events(const OSSL_PARAM params[], void *arg)
+{
+    const OSSL_PARAM *p = NULL;
+    const char *phase = NULL, *type = NULL, *desc = NULL;
+    int ret = 0;
```
Comment (jallisonciq) [inline apps/openssl.c]: The ret variable here doesn't actually do anything. It isn't referenced in anything except "goto err" (where 0 is returned) and set to 1 at the 'end:' label.

You could remove it, and just change the end of the function to be:
```
end:
    return 1;
err:
    return 0;
```

In fact you *could* just replace all 'goto err;' calls with 'return 0' and replace 'goto end' with 'return 1' as you're not doing cleanup in any of those code paths.

If you think you might want change this to do resource cleanup later, then feel free to ignore the last suggestion.

**Pull request merged**